# HG changeset patch
# User Rik <rik@octave.org>
# Date 1489371147 25200
#      Sun Mar 12 19:12:27 2017 -0700
# Node ID 79122fab82ee9cce288257ce143767f5a406c90f
# Parent  a921b6b772f3a8c01a07cb1bb2f708e874fc7649
Restore all figure settings after clf ("reset") (bug #50527).

* scripts/plot/util/private/__set_default_mouse_modes__.m: New script extracted
from figure.m

* scripts/plot/util/module.mk: Add script to build system.

* figure.m: Remove local function __add_default_mouse_modes__ and call
__set_default_mouse_modes__ instead.

* clf.m: When doing a full reset, call __add_default_menu__ and
__set_default_mouse_modes__ to restore these settings.

diff --git a/scripts/plot/util/clf.m b/scripts/plot/util/clf.m
--- a/scripts/plot/util/clf.m
+++ b/scripts/plot/util/clf.m
@@ -68,16 +68,18 @@ function h = clf (varargin)
       print_usage ();
     endif
   endif
 
   if (do_reset)
     ## Select all the children, including the one with hidden handles.
     delete (allchild (hfig));
     reset (hfig);
+    __add_default_menu__ (hfig);
+    __set_default_mouse_modes__ (hfig);
   else
     ## Select only the chilren with visible handles.
     delete (get (hfig, "children"));
 
     ## Also delete the annotation axes
     hover = findall (hfig, "-depth", 1, "tag", "scribeoverlay");
     delete (hover);
   endif
diff --git a/scripts/plot/util/figure.m b/scripts/plot/util/figure.m
--- a/scripts/plot/util/figure.m
+++ b/scripts/plot/util/figure.m
@@ -80,52 +80,35 @@ function h = figure (varargin)
     if (init_new_figure || cf != f)
       drawnow ();
     endif
   endif
 
   if (init_new_figure)
     f = __go_figure__ (f, varargin{:});
     __add_default_menu__ (f);
-    __add_default_mouse_modes__ (f);
+    __set_default_mouse_modes__ (f);
   elseif (nargs > 0)
     set (f, varargin{:});
   endif
 
   set (0, "currentfigure", f);
   ## When switching to figure N, make figure visible and on top of stack,
-  ## unless visibility is explicitly switched off
+  ## unless visibility is explicitly switched off.
   if (! init_new_figure && ! any (strcmpi (varargin(1:2:end), "visible")
                                   && strcmpi (varargin(2:2:end), "off")))
     set (f, "visible", "on");
   endif
 
   if (nargout > 0)
     h = f;
   endif
 
 endfunction
 
-function __add_default_mouse_modes__ (fig)
-
-  set (fig, "__pan_mode__", struct ("Enable", "off",
-                                    "Motion", "both",
-                                    "FigureHandle", fig));
-
-  set (fig, "__rotate_mode__", struct ("Enable", "off",
-                                       "RotateStyle", "box",
-                                       "FigureHandle", fig));
-
-  set (fig, "__zoom_mode__", struct ("Enable", "off",
-                                     "Motion", "both",
-                                     "Direction", "in",
-                                     "FigureHandle", fig));
-
-endfunction
-
 
 %!test
 %! hf = figure ("visible", "off");
 %! unwind_protect
 %!   assert (hf, gcf);
 %!   assert (isfigure (hf));
 %!   hf2 = figure (hf, "visible", "off");
 %!   assert (hf, hf2);
diff --git a/scripts/plot/util/module.mk b/scripts/plot/util/module.mk
--- a/scripts/plot/util/module.mk
+++ b/scripts/plot/util/module.mk
@@ -15,17 +15,18 @@ scripts_plot_util_PRIVATE_FCN_FILES = \
   scripts/plot/util/private/__gnuplot_has_feature__.m \
   scripts/plot/util/private/__gnuplot_has_terminal__.m \
   scripts/plot/util/private/__gnuplot_open_stream__.m \
   scripts/plot/util/private/__gnuplot_print__.m \
   scripts/plot/util/private/__gnuplot_version__.m \
   scripts/plot/util/private/__gnuplot_draw_axes__.m \
   scripts/plot/util/private/__gnuplot_draw_figure__.m \
   scripts/plot/util/private/__opengl_print__.m \
-  scripts/plot/util/private/__print_parse_opts__.m
+  scripts/plot/util/private/__print_parse_opts__.m \
+  scripts/plot/util/private/__set_default_mouse_modes__.m
 
 scripts_plot_util_FCN_FILES = \
   scripts/plot/util/__actual_axis_position__.m \
   scripts/plot/util/__default_plot_options__.m \
   scripts/plot/util/__gnuplot_drawnow__.m \
   scripts/plot/util/__next_line_color__.m \
   scripts/plot/util/__next_line_style__.m \
   scripts/plot/util/__opengl_info__.m \
diff --git a/scripts/plot/util/private/__set_default_mouse_modes__.m b/scripts/plot/util/private/__set_default_mouse_modes__.m
new file mode 100644
--- /dev/null
+++ b/scripts/plot/util/private/__set_default_mouse_modes__.m
@@ -0,0 +1,38 @@
+## Copyright (C) 2017 Rik Wehbring
+##
+## This file is part of Octave.
+##
+## Octave is free software; you can redistribute it and/or modify it
+## under the terms of the GNU General Public License as published by
+## the Free Software Foundation; either version 3 of the License, or
+## (at your option) any later version.
+##
+## Octave is distributed in the hope that it will be useful, but
+## WITHOUT ANY WARRANTY; without even the implied warranty of
+## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+## GNU General Public License for more details.
+##
+## You should have received a copy of the GNU General Public License
+## along with Octave; see the file COPYING.  If not, see
+## <http://www.gnu.org/licenses/>.
+
+## -*- texinfo -*-
+## @deftypefn {} {} __set_default_mouse_modes__.m (@var{fig})
+## Set mouse mode properties of figure to default values.
+##
+## @end deftypefn
+
+function __set_default_mouse_modes__ (fig)
+
+  set (fig, "__pan_mode__", struct ("Enable", "off",
+                                    "Motion", "both",
+                                    "FigureHandle", fig),
+            "__rotate_mode__", struct ("Enable", "off",
+                                       "RotateStyle", "box",
+                                       "FigureHandle", fig),
+            "__zoom_mode__", struct ("Enable", "off",
+                                     "Motion", "both",
+                                     "Direction", "in",
+                                     "FigureHandle", fig));
+
+endfunction
